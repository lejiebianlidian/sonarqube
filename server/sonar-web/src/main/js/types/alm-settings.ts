/*
 * SonarQube
 * Copyright (C) 2009-2020 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
export const enum AlmKeys {
  Azure = 'azure',
  Bitbucket = 'bitbucket',
  GitHub = 'github',
  GitLab = 'gitlab'
}

export interface AlmBindingDefinition {
  key: string;
}

export interface AzureBindingDefinition extends AlmBindingDefinition {
  personalAccessToken: string;
}

export interface BitbucketBindingDefinition extends AlmBindingDefinition {
  personalAccessToken: string;
  url: string;
}

export interface GithubBindingDefinition extends AlmBindingDefinition {
  appId: string;
  clientId: string;
  clientSecret: string;
  privateKey: string;
  url: string;
}

export interface GitlabBindingDefinition extends AlmBindingDefinition {
  personalAccessToken: string;
  url?: string;
}

export interface ProjectAlmBindingResponse {
  alm: AlmKeys;
  key: string;
  repository?: string;
  slug?: string;
  summaryCommentEnabled?: boolean;
}

export interface ProjectBitbucketBindingResponse extends ProjectAlmBindingResponse {
  alm: AlmKeys.Bitbucket;
  repository: string;
  slug: string;
}

export interface ProjectAlmBindingParams {
  almSetting: string;
  project: string;
}

export interface AzureProjectAlmBindingParams extends ProjectAlmBindingParams {}

export interface BitbucketProjectAlmBindingParams extends ProjectAlmBindingParams {
  repository: string;
  slug: string;
}

export interface GithubProjectAlmBindingParams extends ProjectAlmBindingParams {
  repository: string;
  summaryCommentEnabled: boolean;
}

export interface GitlabProjectAlmBindingParams extends ProjectAlmBindingParams {
  repository?: string;
}

export interface AlmSettingsInstance {
  alm: AlmKeys;
  key: string;
  url?: string;
}

export interface AlmSettingsBindingDefinitions {
  [AlmKeys.Azure]: AzureBindingDefinition[];
  [AlmKeys.Bitbucket]: BitbucketBindingDefinition[];
  [AlmKeys.GitHub]: GithubBindingDefinition[];
  [AlmKeys.GitLab]: GitlabBindingDefinition[];
}
