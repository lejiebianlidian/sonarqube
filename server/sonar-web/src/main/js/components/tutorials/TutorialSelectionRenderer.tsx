/*
 * SonarQube
 * Copyright (C) 2009-2020 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */

import * as React from 'react';
import { translate } from 'sonar-ui-common/helpers/l10n';
import { getBaseUrl } from 'sonar-ui-common/helpers/urls';
import { AlmBindingDefinition, ProjectAlmBindingResponse } from '../../types/alm-settings';
import JenkinsTutorial from './jenkins/JenkinsTutorial';
import ManualTutorial from './manual/ManualTutorial';
import { TutorialModes } from './types';

export interface TutorialSelectionRendererProps {
  almBinding?: AlmBindingDefinition;
  component: T.Component;
  currentUser: T.LoggedInUser;
  loading: boolean;
  onSelectTutorial: (mode: TutorialModes) => void;
  projectBinding?: ProjectAlmBindingResponse;
  selectedTutorial?: TutorialModes;
}

export default function TutorialSelectionRenderer(props: TutorialSelectionRendererProps) {
  const { almBinding, component, currentUser, loading, projectBinding, selectedTutorial } = props;

  if (loading) {
    return <i className="spinner" />;
  }

  return (
    <>
      {selectedTutorial === undefined && (
        <div className="tutorial-selection">
          <header className="spacer-top spacer-bottom padded">
            <h1 className="text-center big-spacer-bottom">
              {translate('onboarding.tutorial.choose_method')}
            </h1>
          </header>

          <div className="display-flex-space-around">
            <button
              className="button button-huge display-flex-column tutorial-mode-jenkins"
              onClick={() => props.onSelectTutorial(TutorialModes.Jenkins)}
              type="button">
              <img
                alt="" // Should be ignored by screen readers
                height={80}
                src={`${getBaseUrl()}/images/tutorials/jenkins.svg`}
              />
              <div className="medium big-spacer-top">
                {translate('onboarding.tutorial.choose_method.jenkins')}
              </div>
            </button>

            <button
              className="button button-huge display-flex-column tutorial-mode-manual"
              onClick={() => props.onSelectTutorial(TutorialModes.Manual)}
              type="button">
              <img
                alt="" // Should be ignored by screen readers
                height={80}
                src={`${getBaseUrl()}/images/sonarcloud/analysis/manual.svg`}
              />
              <div className="medium big-spacer-top">
                {translate('onboarding.tutorial.choose_method.manual')}
              </div>
            </button>
          </div>
        </div>
      )}

      {selectedTutorial === TutorialModes.Manual && (
        <ManualTutorial component={component} currentUser={currentUser} />
      )}

      {selectedTutorial === TutorialModes.Jenkins && projectBinding !== undefined && (
        <JenkinsTutorial
          almBinding={almBinding}
          component={component}
          projectBinding={projectBinding}
        />
      )}
    </>
  );
}
