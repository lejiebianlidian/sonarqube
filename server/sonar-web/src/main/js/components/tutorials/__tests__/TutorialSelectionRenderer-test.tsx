/*
 * SonarQube
 * Copyright (C) 2009-2020 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */

import { shallow } from 'enzyme';
import * as React from 'react';
import { click } from 'sonar-ui-common/helpers/testUtils';
import {
  mockBitbucketBindingDefinition,
  mockProjectBitbucketBindingGet
} from '../../../helpers/mocks/alm-settings';
import { mockComponent, mockLoggedInUser } from '../../../helpers/testMocks';
import TutorialSelectionRenderer, {
  TutorialSelectionRendererProps
} from '../TutorialSelectionRenderer';
import { TutorialModes } from '../types';

it('should render correctly', () => {
  expect(shallowRender()).toMatchSnapshot('selection');
  expect(shallowRender({ loading: true })).toMatchSnapshot('loading');
  expect(shallowRender({ selectedTutorial: TutorialModes.Manual })).toMatchSnapshot(
    'manual tutorial'
  );
  expect(
    shallowRender({
      selectedTutorial: TutorialModes.Jenkins,
      projectBinding: mockProjectBitbucketBindingGet()
    })
  ).toMatchSnapshot('jenkins tutorial');
});

it('should allow mode selection', () => {
  const onSelectTutorial = jest.fn();
  const wrapper = shallowRender({ onSelectTutorial });

  click(wrapper.find('button.tutorial-mode-jenkins'));
  expect(onSelectTutorial).toHaveBeenLastCalledWith(TutorialModes.Jenkins);

  click(wrapper.find('button.tutorial-mode-manual'));
  expect(onSelectTutorial).toHaveBeenLastCalledWith(TutorialModes.Manual);
});

function shallowRender(props: Partial<TutorialSelectionRendererProps> = {}) {
  return shallow<TutorialSelectionRendererProps>(
    <TutorialSelectionRenderer
      almBinding={mockBitbucketBindingDefinition()}
      component={mockComponent()}
      currentUser={mockLoggedInUser()}
      loading={false}
      onSelectTutorial={jest.fn()}
      {...props}
    />
  );
}
