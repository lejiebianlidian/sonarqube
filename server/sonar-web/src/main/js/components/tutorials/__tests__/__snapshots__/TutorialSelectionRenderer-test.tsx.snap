// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render correctly: jenkins tutorial 1`] = `
<Fragment>
  <Connect(JenkinsTutorial)
    almBinding={
      Object {
        "key": "key",
        "personalAccessToken": "asdf1234",
        "url": "http://bbs.enterprise.com",
      }
    }
    component={
      Object {
        "breadcrumbs": Array [],
        "key": "my-project",
        "name": "MyProject",
        "organization": "foo",
        "qualifier": "TRK",
        "qualityGate": Object {
          "isDefault": true,
          "key": "30",
          "name": "Sonar way",
        },
        "qualityProfiles": Array [
          Object {
            "deleted": false,
            "key": "my-qp",
            "language": "ts",
            "name": "Sonar way",
          },
        ],
        "tags": Array [],
      }
    }
    projectBinding={
      Object {
        "alm": "bitbucket",
        "key": "foo",
        "repository": "PROJECT_KEY",
        "slug": "repo-slug",
      }
    }
  />
</Fragment>
`;

exports[`should render correctly: loading 1`] = `
<i
  className="spinner"
/>
`;

exports[`should render correctly: manual tutorial 1`] = `
<Fragment>
  <ManualTutorial
    component={
      Object {
        "breadcrumbs": Array [],
        "key": "my-project",
        "name": "MyProject",
        "organization": "foo",
        "qualifier": "TRK",
        "qualityGate": Object {
          "isDefault": true,
          "key": "30",
          "name": "Sonar way",
        },
        "qualityProfiles": Array [
          Object {
            "deleted": false,
            "key": "my-qp",
            "language": "ts",
            "name": "Sonar way",
          },
        ],
        "tags": Array [],
      }
    }
    currentUser={
      Object {
        "groups": Array [],
        "isLoggedIn": true,
        "login": "luke",
        "name": "Skywalker",
        "scmAccounts": Array [],
      }
    }
  />
</Fragment>
`;

exports[`should render correctly: selection 1`] = `
<Fragment>
  <div
    className="tutorial-selection"
  >
    <header
      className="spacer-top spacer-bottom padded"
    >
      <h1
        className="text-center big-spacer-bottom"
      >
        onboarding.tutorial.choose_method
      </h1>
    </header>
    <div
      className="display-flex-space-around"
    >
      <button
        className="button button-huge display-flex-column tutorial-mode-jenkins"
        onClick={[Function]}
        type="button"
      >
        <img
          alt=""
          height={80}
          src="/images/tutorials/jenkins.svg"
        />
        <div
          className="medium big-spacer-top"
        >
          onboarding.tutorial.choose_method.jenkins
        </div>
      </button>
      <button
        className="button button-huge display-flex-column tutorial-mode-manual"
        onClick={[Function]}
        type="button"
      >
        <img
          alt=""
          height={80}
          src="/images/sonarcloud/analysis/manual.svg"
        />
        <div
          className="medium big-spacer-top"
        >
          onboarding.tutorial.choose_method.manual
        </div>
      </button>
    </div>
  </div>
</Fragment>
`;
